flink-runtime/src/main/java/org/apache/flink/runtime/io/disk/FileChannelManagerImpl.java

```diff
@@ -51,12 +52,30 @@
 	/** The number of the next path to use. */
 	private volatile int nextPath;
 
+	private final String prefix;
+
+	private final Thread shutdownHook;
+
 	public FileChannelManagerImpl(String[] tempDirs, String prefix) {
+		this(tempDirs, prefix, false);
+	}
+
+	public FileChannelManagerImpl(String[] tempDirs, String prefix, boolean deleteOnShutdown) {
 		checkNotNull(tempDirs, "The temporary directories must not be null.");
 		checkArgument(tempDirs.length > 0, "The temporary directories must not be empty.");
 
 		this.random = new Random();
 		this.nextPath = 0;
+		this.prefix = prefix;
+
+		if (deleteOnShutdown) {
+			shutdownHook = ShutdownHookUtil.addShutdownHook(this, String.format("%s-%s", getClass().getSimpleName(), prefix), LOG);
```

Comment:
Can we always add the shutdown hook to remove the argument `deleteOnShutdown`?

Reply:
This is due to in this PR I think it should be better to not affect the current behaviors of `IOManager`.  Although `IOManager` also have the problem that register shutdown hook after the directories are created, I think it might be better to fix it in separate PR if we agree with how to deal with it. Since `IOManager` already has a shutdown hook, for `IOManager` we also need to allow the `FileChannelManagerImpl#close` be closed for multiple times (i.e., using `closed` flag to skip cleanup in repeat `close` calls).

Reply:
I got the point now. 

I think it is better to make `FileChannelManagerImpl` always have the `shutdownHook`, and then it does not need judge whether it is `null` in other usages. Also it does not need to care about this argument `deleteOnShutdown` to true or false for upper component. 

In order to avoid repeated `close` calling by `IOManager` usage, we can add the atomic boolean var in `FileChannelManagerImpl` and check this state in other usages as well.